Examples/Io/Root/src/RootTrackParameterWriter.cpp

```diff
@@ -63,30 +114,106 @@ ActsExamples::ProcessCode ActsExamples::RootTrackParameterWriter::endRun() {
 
   // Exclusive access to the tree while writing
   std::lock_guard<std::mutex> lock(m_writeMutex);
 
   // Get the event number
   m_eventNr = ctx.eventNumber;
 
-  for (auto& params : trackParams) {
-    m_d0 = params.parameters()[0];
-    m_z0 = params.parameters()[1];
-    m_phi = params.parameters()[2];
-    m_theta = params.parameters()[3];
-    m_qp = params.parameters()[4];
+  // Loop over the estimated track parameters
+  for (size_t iparams = 0; iparams < trackParams.size(); ++iparams) {
+    // The reference surface of the parameters, i.e. also the reference surface
+    // of the first space point
+    const auto& surface = trackParams[iparams].referenceSurface();
+    // The estimated bound parameters vector
+    const auto params = trackParams[iparams].parameters();
+    m_loc0 = params[Acts::eBoundLoc0];
+    m_loc1 = params[Acts::eBoundLoc1];
+    m_phi = params[Acts::eBoundPhi];
+    m_theta = params[Acts::eBoundTheta];
+    m_qop = params[Acts::eBoundQOverP];
+    m_time = params[Acts::eBoundTime];
+    m_p = std::abs(1.0 / m_qop);
+    m_pt = m_p * std::sin(m_theta);
+    m_eta = std::atanh(std::cos(m_theta));
+
+    // Get the seed via the estimated parameters to seed map
+    const auto& groupedSeedIdx = trackParametersSeedMap.at(iparams);
+    const auto& seed = seeds[groupedSeedIdx.regionIdx][groupedSeedIdx.seedIdx];
+    // check the seed quality
+    ProtoTrack ptrack{seed.sp()[0]->measurementIndex(),
+                      seed.sp()[1]->measurementIndex(),
+                      seed.sp()[2]->measurementIndex()};
+    std::vector<ParticleHitCount> particleHitCounts;
+    identifyContributingParticles(hitParticlesMap, ptrack, particleHitCounts);
+    m_truthMatched = false;
+    if (particleHitCounts.size() == 1) {
+      m_truthMatched = true;
+    }
+    // Get the index of the first space point
+    const auto& firstSP = seed.sp().front();
+    const auto& hitIdx = firstSP->measurementIndex();
+    // Get the sim hits via the measurement to sim hits map
+    auto indices = makeRange(hitSimHitsMap.equal_range(hitIdx));
+    auto [truthLocal, truthPos4, truthUnitDir] =
+        detail::averageSimHits(ctx.geoContext, surface, simHits, indices);
+    // Get the truth track parameter at the first space point
+    m_t_loc0 = truthLocal[Acts::ePos0];
+    m_t_loc1 = truthLocal[Acts::ePos1];
+    m_t_phi = phi(truthUnitDir);
+    m_t_theta = theta(truthUnitDir);
+    m_t_time = truthPos4[Acts::eTime];
+    // momemtum averaging makes even less sense than averaging position and
+    // direction. use the first momentum or set q/p to zero
+    if (not indices.empty()) {
+      // we assume that the indices are within valid ranges so we do not
+      // need to check their validity again.
+      const auto simHitIdx0 = indices.begin()->second;
+      const auto& simHit0 = *simHits.nth(simHitIdx0);
+      const auto p =
+          simHit0.momentum4Before().template segment<3>(Acts::eMom0).norm();
```

Comment:
Where does `momentum4Before` come from? Could the name be changed? (Obviously only if this is part of the changeset somehow.)

Reply:
The `momentum4Before` is a member function of the Fatras `Hit`. It meas the four-momentum before the material effects.